### PR TITLE
Add VQVAE2 with PixelCNN generators to recognition models

### PR DESCRIPTION
**Pull request by leobaldock**
This PR implements the network VQVAE2 and accompanying PixelCNN generative models for 2D images. The models are implemented in PyTorch and pretrained states are provided for the models trained on 2D slices from the OASIS MRI dataset. Scripts are also provided to train the models on the OASIS dataset, along with a script to extract the slices from the 3D images. The evaluation notebook can be used to test the trained networks on the OASIS dataset. 

The included files are as follows:
 - **vqvae.py**: VQVAE2 model implementation.
 - **pixel_cnn**: Folder containing PixelCNN models.
     - **attention.py**: Implementation of self-attention layer.
     - **pixel_cnn.py**: PixelCNN layer implementations.
     - **prior_models.py**: PixelCNN model architectures for top- and bottom-level generators.
 - **results/oasis**: Folder containing pretrained model states.
 - **evaluate.ipynb**: Notebook for testing trained networks.
 - **train_vqvae.py**: Script to train the VQVAE2 model on the OASIS dataset.
 - **train_pixel_cnn.py**: Script to train the PixelCNN models on the OASIS dataset.
 - **extract_slices.py**: Script to build the 2D dataset from OASIS T1W scans.
 - **load_dataset.py**: Methods for loading the 2D slices using PyTorch Datasets.
 - **ssim.py**: Implementation of the SSIM metric.
 - **trainer.py**: An generic PyTorch model trainer class.

Leo Baldock
44310776

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
Code
- Strcucture OK
- Commenting OK
- Please do not commit model state files (.pth)

README 
- Please fix broken image paths in read me 
- Content OK

General Comments
- Commit messages OK

**Comment by shakes76** [comment]
## TF/Torch Usage
Adequate use and implementation
## Good Practice (Design/Commenting)
Good spacing and comments
Header blocks
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage 
Module present 
No Data leakage 
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
Please remove model files, these will clog up the repository. Your pull request will not be merged until they are removed. Feedback marks will be awarded when these are removed. -2
Request Description good

**Comment by leobaldock** [comment]
Thanks Shakes, thanks Siyu, I've made those changes.

**Comment by shakes76** [comment]
Updated with feedback marks.
